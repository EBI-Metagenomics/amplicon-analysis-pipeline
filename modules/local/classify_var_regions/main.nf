
process CLASSIFY_VAR_REGIONS {
    tag "$meta.id"
    label 'light'
    conda "${projectDir}/conf/environment.yml"

    input:
    tuple val(meta), path(cmsearch_deoverlap_out)

    output:
    tuple val(meta), path ("*S.V*.txt", arity: '0..2'), optional: true, emit: classify_var_regions
    tuple val(meta), val("concat"), path("*.concat.regions.txt"), optional: true, emit: concat_var_regions
    tuple val(meta), path("*.tsv"), optional: true, emit: cdch_out

    """
    classify_var_regions -d ./ -o ${meta.id} --statistics $cmsearch_deoverlap_out
    
    num_files="\$(ls *S.V*.txt | wc -l  )"
    if [ \$num_files -gt 1 ]; then
        cat *S.V*.txt > ${meta.id}.concat.regions.txt
    fi
    """

}
