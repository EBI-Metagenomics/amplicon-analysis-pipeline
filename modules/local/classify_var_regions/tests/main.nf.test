nextflow_process {

    name "Test Process CLASSIFY_VAR_REGIONS"
    script "../main.nf"
    process "CLASSIFY_VAR_REGIONS"
    tag "modules"
    name "classify_var_regions"

    test("Should take a .tblout.deoverlapped file and generate a tsv file for submitting metadata to CDCH, except it will only have the header line due to insufficient read count") {
        tag "classify_var_regions_failed"
        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('${projectDir}/modules/local/classify_var_regions/tests/data/test_insufficient_reads.tblout.deoverlapped', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.cdch_out).match("cdch_out_failed") }
            )
        }

    }

    test("Should take a .tblout.deoverlapped file and generate two output files - a tsv file with actual data, and a file containing read headers for the identified var_region") {
        tag "classify_var_regions_success_single"

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('${projectDir}/modules/local/classify_var_regions/tests/data/test_success_single.tblout.deoverlapped', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.classify_var_regions).match("classify_var_regions_single") },
                { assert snapshot(process.out.cdch_out).match("cdch_out_success_single") }
            )
        }

    }

    test("Should take a .tblout.deoverlapped file and generate three output files - a tsv file with actual data, a list of files containing read headers for the identified var_regions, and a final concatenated read header file") {
        tag "classify_var_regions_success_double"

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('${projectDir}/modules/local/classify_var_regions/tests/data/test_success_double.tblout.deoverlapped', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.classify_var_regions).match("classify_var_regions_double") },
                { assert snapshot(process.out.cdch_out).match("cdch_out_success_double") },
                { assert snapshot(process.out.cdch_out).match("classify_var_regions_concat") }

            )
        }

    }

}
