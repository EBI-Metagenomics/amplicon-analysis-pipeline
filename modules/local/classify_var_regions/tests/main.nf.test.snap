{
    "classify_var_regions_concat": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,601bd9e9445563f88c0c79bedbd0adaf"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:21.142336898"
    },
    "cdch_out_success_double": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,601bd9e9445563f88c0c79bedbd0adaf"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:21.106948756"
    },
    "classify_var_regions_single": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test.16S.V3-V4.txt:md5,26b4cbbafdcfba30abb7ac561be0a40e"
                    ]
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:11.755494749"
    },
    "cdch_out_failed": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,f0abe32d6c212b764bc9b04ba3f69dab"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:02.025864489"
    },
    "cdch_out_success_single": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.tsv:md5,aba8ff1cc6df1ce076633dd844b95d28"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:11.784802851"
    },
    "classify_var_regions_double": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test.16S.V3-V4.txt:md5,26b4cbbafdcfba30abb7ac561be0a40e",
                        "test.18S.V9.txt:md5,eaf0569f6d4841da453b733f820bff6b"
                    ]
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:54:21.06575426"
    }
}